### PR TITLE
Auth0 !

### LINKED ISSUE #42: Authentication Loop
Sometimes, with some users, with some browsers, there is a authentication loop trough Google Auth.
It will be nice to migrate to passport+google auth to Auth0 services for the users authentication.
Apparently, this only happens on Firefox.

So for now, we will step by step rebuild a new zenika resume App.

### PR DESCRIPTION
**Pull request by hgwood**
This replaces the Google Sign-in authentication by Auth0. 

Why?
- Easier to manage authentication processes across all apps
- Should fix #42
- Authentication code on the back-end was really hard to follow and was using non-production modes
- The currently used Google Sign-in API is deprecated and will no longer work starting from March 7 (see https://github.com/Zenika/dreamlab/issues/109)

:warning: Must not be merged before we add an exception to https://github.com/Zenika/auth0/blob/master/rules/only-allow-zenika-domain.js to allow zenika.resume@gmail.com.

ℹ️ one #44 is merged this should be rebased on master without https://github.com/Zenika/Zenika-Resume/pull/43/commits/2a93561b7a62472a05bbdcbc147427d549f6d05c in it

Done in tandem with @ClementVanPeuter

### CODE REVIEW

File: app/auth.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import auth0 from 'auth0-js';
+class Auth {
+  constructor() {
+    this.auth0 = new auth0.WebAuth({
+      domain: 'zenika.eu.auth0.com',
+      clientID: '60hXTJPUSI2lD7gMilLEvOa7DH1zC0WO',
+      redirectUri: 'http://localhost:3000/',
```
Comment (hgwood) [inline app/auth.js]: Need to change this.

File: app/auth.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    localStorage.setItem('isLoggedIn', 'true');
+
+    // Set the time that the access token will expire at
+    let expiresAt = (authResult.expiresIn * 1000) + new Date().getTime();
+    this.accessToken = authResult.accessToken;
+    this.idToken = authResult.idToken;
+    this.expiresAt = expiresAt;
+  }
+
+  renewSession() {
```
Comment (hgwood) [inline app/auth.js]: Either use this or remove it.

File: app/auth.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    // access token's expiry time
+    let expiresAt = this.expiresAt;
+    return new Date().getTime() < expiresAt;
+  }
+
+}
+
+const auth = new Auth();
+
+export const redirectToLogin = () => {
```
Comment (hgwood) [inline app/auth.js]: I don't think this is used anymore.

File: app/components/App.jsx [hunk 4]
```diff
@@ -114,8 +116,8 @@ class App extends Component {
       });
     });
 
-    this.props.controller.on(Events.AUTHENTICATION_REQUIRED, (state) => {
-      document.location.href = '/login/google?uuid=' + window.location.pathname.slice(1);
+    this.props.controller.on(Events.AUTHENTICATION_REQUIRED, () => {
+      redirectToLogin();
```
Comment (hgwood) [inline app/components/App.jsx]: Can be replaced with `auth.login()`

File: app/index.jsx [hunk 5]
```diff
@@ -39,17 +40,31 @@ const theme = createMuiTheme({
...
+      appElement
+    );
+  } else {
+    auth.login();
+  }
+}).catch(err => {
+  // eslint-disable-next-line no-console
+  console.error(err);
+  // eslint-disable-next-line no-alert
+  alert('Authentication error! Check your console for more details.');
```
Comment (hgwood) [inline app/index.jsx]: More detailed message needed. "try again, if problem persists contact dreamlab ..."

File: server.js [hunk 6]
```diff
@@ -162,67 +96,19 @@ function executeQueryWithCallback(query, params, response, callback) {
...
-);
-
-app.get('/me', function (req, res) {
-  if (!isUserConnectedAndZenika(req)) {
-    res.redirect('/login/google');
-    return;
-  }
-  res.status(200).json(req.user);
+app.get('/me', jwtCheck, function (req, res) {
+  res.status(200).json({ photos: [{ value: 'https://images.pexels.com/photos/20787/pexels-photo.jpg' }] });
```
Comment (hgwood) [inline server.js]: Need to implement this more seriously :p

File: server.js [hunk 7]
```diff
@@ -162,67 +96,19 @@ function executeQueryWithCallback(query, params, response, callback) {
...
-      res.redirect('/' + req.session.requestedUuid);
-    } else {
-      res.redirect('/');
-    }
-  }
-);
-
 // Match UUIDs
-app.get('/[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}', ensureAuthenticated, (req, res) => {
+app.get('/[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}', (req, res) => {
```
Comment (hgwood) [inline server.js]: Is this still needed?

File: server.js [hunk 8]
```diff
@@ -369,15 +241,10 @@ api.get('/resumes', (req, res) => {
...
-  if (!isUserConnectedAndZenika(req) || !req.user.emails[0].value) {
-    res.status(401).json();
-    return;
-  }
-
+api.get('/resumes/mine', jwtCheck, (req, res) => {
   executeQueryWithCallback(
     'SELECT uuid, metadata, path, version, last_modified FROM resume WHERE metadata LIKE $1 ORDER BY last_modified DESC',
-    [`%${req.user.emails[0].value}%`],
+    [`%hugo.wood@zenika.com%`],
```
Comment (hgwood) [inline server.js]: Need to implement this

**Comment by Errorname** [comment]
Instead of adding a rule for "zenika.resume@gmail.com", could we not create a "resume@zenika.com" ?

**Comment by hgwood** [comment]
No because that would give people access to too much.
